### PR TITLE
Force semantic commands to use command

### PR DESCRIPTION
**Pull request by fcollonval**
This PR enforces commands as the main way to define actions even for semantic commands. It also allow to always call the same basic command for semantic commands composing multiple basic commands to ensure homogeneous behaviour.

## Code changes

Semantic commands reuse information from underlying commands.

## User-facing changes

Small changes in the labels as they are built from the underlying commands

## Backwards-incompatible changes

Breaks the semantic commands API.

### CODE REVIEW

**Comment by fcollonval** [comment]
@jess-x @3coins @ohrely could one (or more :wink:) of you review this PR please?

**Comment by 3coins** [comment]
@fcollonval 
I will take a look over the weekend.

**Comment by jess-x** [comment]
@fcollonval let me also try and take a look

**Review by 3coins — COMMENTED** [review]
Does this impact any external extensions?

File: packages/application/src/utils.ts [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+ */
+export function createSemanticCommand(
+  app: JupyterFrontEnd,
+  semanticCommands: SemanticCommand | SemanticCommand[],
+  defaultValues: ISemanticCommandDefault,
+  trans: TranslationBundle
+): CommandRegistry.ICommandOptions {
+  const { commands, shell } = app;
+  const commandList =
+    semanticCommands instanceof Array ? semanticCommands : [semanticCommands];
```
[the remark this answers is not in the document]
Comment (fcollonval) [reply packages/application/src/utils.ts:51]: Very good point

File: packages/application/src/utils.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+          const end = labels.slice(-1)[0] + (hasEllipsis ? '…' : '');
+          return trans.__('%1 and %2', main, end);
+        }
+      }
+    },
+    caption: () => {
+      const widget = shell.currentWidget;
+      const commandIds = commandList.map(cmd =>
+        widget !== null ? cmd.getActiveCommandId(widget) : null
+      );
```
[the remark this answers is not in the document]
Comment (fcollonval) [reply packages/application/src/utils.ts:82]: :+1:

File: packages/help-extension/src/index.tsx [hunk 3]
```diff
@@ -302,6 +298,41 @@ const resources: JupyterFrontEndPlugin<void> = {
       return widget;
     }
 
+    commands.addCommand(CommandIDs.open, {
+      label: args => args['text'] as string,
+      execute: args => {
+        const url = args['url'] as string;
+        const text = args['text'] as string;
+        const newBrowserTab = (args['newBrowserTab'] as boolean) || false;
```
[the remark this answers is not in the document]
Comment (fcollonval) [reply packages/help-extension/src/index.tsx:306]: No you cannot. what could be done is `const {url, text, newBrowserTab} = args as {url: string, text: string, newBrowserTab: boolean};` but you cannot set default value.

**Comment by fcollonval** [comment]
Thanks for the review @3coins I applied the suggestions and added some tests. If you could have a second look, it will be appreciated.

> Does this impact any external extensions?

I don't know of an extension using it. But it will impact external extensions using those API (hence the api-changes tag and the note in the PR description).

File: packages/help-extension/src/index.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (3coins) [reply packages/help-extension/src/index.tsx:306]: Thanks for explaining. Looks good.

**Review by 3coins — APPROVED** [review]
Thanks for making changes. LGTM!

**Comment by jess-x** [comment]
I learned a lot trying to make sense of this PR. Thanks for pinging me on this one! @fcollonval

**Comment by jtpio** [comment]
> Breaks the semantic commands API.

Should the affected packages have their major version bumped?

**Pull request merged**